platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/counters/BackpressureObjectCounter.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.wiring.counters;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.LongUnaryOperator;
+
+/**
+ * A utility for counting the number of objects in a various part of the pipeline. Will apply backpressure if the number
+ * of objects exceeds a specified capacity.
+ */
+public class BackpressureObjectCounter extends ObjectCounter {
+
+    private final AtomicLong count = new AtomicLong(0);
+    private final LongUnaryOperator increment;
+    private final long sleepNanos;
+
+    private static class NoCapacityException extends RuntimeException {
+        public NoCapacityException() {}
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param capacity      the maximum number of objects that can be in the part of the system that this object is
+     *                      being used to monitor before backpressure is applied
+     * @param sleepDuration when there is no capacity available, sleep for this duration before trying again, if null
+     *                      then do not sleep
+     */
+    public BackpressureObjectCounter(final long capacity, @Nullable Duration sleepDuration) {
+
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("Capacity must be greater than zero");
+        }
+
+        if (sleepDuration == null) {
+            sleepNanos = -1;
+        } else {
+            sleepNanos = sleepDuration.toNanos();
+        }
+
+        increment = count -> {
+            if (count >= capacity) {
+                throw new NoCapacityException();
```

Comment:
throwing an exception is expensive, better to use `compareAndSet`

Reply:
I actually think this may be faster in the standard case we want to optimize, i.e. when there is available capacity and we don't need to apply back pressure.

If I use compare and set, I'd have to perform two operations for each on-ramping, even when capacity is not reached.
1) read the current value
2) do a compare and set operation

With the throwing approach, I only have to do one atomic long operation per on-ramping. When we are at capacity there is additional overhead from the exception we are throwing.

In my opinion, we should optimize the performance of the scenario when there is available capacity, even if there are costs for when capacity is scarce.

Reply:
If you look inside of `updateAndGet`, there is a loop, it has to do the same thing basically. So I don't think there is any benefit of this.

You could go with `getAndAccumulate`, that way you only call the method once

Reply:
Throwing an exception is very expensive in performance sensitive code like this (costs CPU and generates a lot of garbage). Especially when it's done in an implicit busy wait loop: `ForkJoinPool.managedBlock()` -> `BackpressureBlocker.isReleasable()`.
We may consider two blocking implementations: busy wait (attempt to increment until successful) and a thread blocking/unblocking one.

Reply:
I went with the two-implementation approach. An exception for the first time we run out of capacity, and compare-and-set for the busy loop.